File: libs/power-log-parser/src/lib/cards/deep-space-curator.ts
import { BlockType, GameTag } from '@firestone-hs/reference-data';
import { Action, ShowEntity } from '../models/action';
import { Tag } from '../models/tag';
import { Node } from '../models/node';
import type { GameState } from '../state/game-state';
import type { StateFacade } from '../state/state-facade';

export class DeepSpaceCurator {
	static GuessTags(
		gameState: GameState,
		creatorCardId: string,
		creatorEntityId: number,
		node: Node,
		stateFacade: StateFacade | null,
	): Tag[] | null {
		if (node.Parent?.Parent?.Type !== Action) {
			return null;
		}

		const act = node.Parent.Parent.Object as Action;
		if (act.Type !== (BlockType.PLAY as number)) {
			return null;
		}

		const playedEntityId = act.Entity;
		const showEntity = act.Data
			.filter((d): d is ShowEntity => d instanceof ShowEntity)
			.find((e) => e.Entity === playedEntityId);

		const guessedTags: Tag[] = [];
		if (showEntity) {
			const tag = new Tag();
			tag.Name = GameTag.COST as number;
			tag.Value = showEntity.GetTag(GameTag.COST);
			guessedTags.push(tag);
		}
		return guessedTags;
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Cards/DeepSpaceCurator.cs
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using HearthstoneReplays.Parser;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Action = HearthstoneReplays.Parser.ReplayData.GameActions.Action;
using HearthstoneReplays.Parser.ReplayData.Entities;

namespace HearthstoneReplays.Events.Cards
{
    internal class DeepSpaceCurator
    {
        public static List<Tag> GuessTags(GameState gameState, string creatorCardId, int creatorEntityId, Node node, StateFacade stateFacade)
        {
            if (node.Parent?.Parent?.Type != typeof(Action))
            {
                return null;
            }
            
            var act = node.Parent.Parent.Object as Action;
            if (act.Type != (int)BlockType.PLAY)
            {
                return null;
            }

            var playedEntityId = act.Entity;
            var showEntity = act.Data
                .Where(d => d is ShowEntity)
                .Select(d => d as ShowEntity)
                .Where(e => e.Entity == playedEntityId)
                .FirstOrDefault();
            var guessedTags = new List<Tag>();
            if (showEntity != null)
            {
                guessedTags.Add(new Tag() { Name = (int)GameTag.COST, Value = showEntity.GetTag(GameTag.COST) });
            }
            return guessedTags;
        }
    }
}
